Trello 144 add a list of most recently annotated by the group urls to the group page

seanh opened the pull request.

seanh commented:
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;Nj6M1pMF&#x2F;144-add-a-list-of-up-to-most-recently-annotated-by-the-group-urls-to-the-group-page"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> Add a list of up to most recently annotated (by the group) URLs to the group page</a>

robertknight commented on h/api/models.py:
```diff
@@ -166,6 +176,76 @@ def title(self):
+
+    @property
+    def document_link(self):
+        """Return a link to this annotation's document.
```
It seems odd to try and render the HTML outside of the template. You could just always render the `<a>` tag in the markup and only include the `href` attribute if a URL is available.

seanh replied:
Yeah, that would simplify things :) Main reason I'm doing this in Python instead of in Jinja2 is that the logic is complex enough that I want unit tests for it. But if we can put the `<a>` tag in either way and the only logic Jinja2 needs is to output the `href` only if there's something to link to, that should be fine

seanh replied:
Ok, having looked into this, I want to keep it how it is. I understand that rendering HTML tags in Python instead of in the template is weird, but:

- This Python function is actually pretty complicated logic, and it's nice to keep it all in one place where it can be tested
- This logic is duplicated in JavaScript code for the annotation cards and now in Python for the page. With the code in Python, we can potentially add the link to the annotation on its way out of our API (in our search lib's `render()` function) and have the client-side code just use it. The problem with this though is annotations that are being created but haven't been saved to the server yet.

If we wanted to move this into the template, I think we'd need four separate Python functions to get the href, title, link_text, and domain, and then template code that looked like this:

    <a {% if href %}href="{{ href }}"{% endif %} title={{ title }}>{{ link_text }}</a>{% if domain %}({{ domain }}){% endif %}

And then even if we added the href, title, link_text, and domain to the annotations in `render()` we'd still have to duplicate that bit of template code containing (admittedly simple) logic client-side.

And splitting this one function into four would not be that simple as the values are inter-dependent (although having four smaller functions would probably benefit the code).

I'm willing to be overridden here though

robertknight commented on h/api/test/models_test.py:
```diff
@@ -78,6 +79,74 @@ def test_title_annotation_that_has_no_document():
+    # If the annotation has an http(s) uri but no document title it should
+    # return the uri linked to the uri.
+    (('http://example.com/example.html', None),
+     '<a href="http://example.com/example.html">http://example.com/example.html</a>'),
```
Given the limited space for the displayed URL, I suggest we should strip uninformative parts such as the scheme and leading 'www.' from what is rendered.

seanh replied:
Good idea!

robertknight commented on h/groups/templates/share.html.jinja2:
```diff
@@ -18,6 +18,18 @@
+        {% endfor %}
+      </ul>
+      {% else %}
+        <p>This group hasn't shared any annotations yet.</p>
```
This sounds slightly odd. The group doesn't share annotations - people in the group do. "No annotations have been shared in this group yet"?

seanh replied:
I think I prefer my wording as it's simpler, and I think users will understand "the group" to mean both the group (i.e. the thing that exists in our database) and the users who're in that group (this seems like typical usage of the word group to me). I just prefer direct language, I'm sick of seeing needlessly indirect things like _this annotation is viewable only by you_ instead of _only you can see this annotation_ etc etc in Hypothesis.

robertknight commented on h/groups/views.py:
```diff
@@ -91,7 +94,35 @@ def _read_group(request, group):
+
+    # TODO: Get shared annotations only.
+    result = search.search(request, params={
+        "group": group.hashid, "limit": 1000})
```
This will need adjusting for Nick's hashid -> pubid conversion

seanh replied:
Done

robertknight commented on h/groups/views.py:
```diff
@@ -91,7 +94,35 @@ def _read_group(request, group):
+    # Group the annotations by URI.
+    annotations_by_uri = collections.OrderedDict()
+    for annotation in annotations:
+        if annotation.uri in annotations_by_uri:
```
Does URI normalization need to be handled here or elsewhere?

seanh replied:
I was thinking that was out of scope for this version, and that this whole thing will work much better once we have annotations and documents in a proper database, but I could probably just pass each uri through our `normalize()` function though

seanh replied:
Done this

seanh commented:
Screenshot of how the page looks currently:

![screenshot from 2015-10-26 11 14 49](https://cloud.githubusercontent.com/assets/22498/10727580/da3cc9f8-7bd2-11e5-806f-e0f110a53d9b.png)

seanh commented:
- [ ] Show the full document title in the tooltip

seanh commented:
This is how this looks currently on production data:

![screenshot from 2015-10-26 17 36 45](https://cloud.githubusercontent.com/assets/22498/10737076/26853b62-7c08-11e5-8c95-d7d5a73a0727.png)

I'm just planning to add some tests now and then this is done.

I added `(hostname|filename)` after each annotation (if it has a hostname or filename that's different from the link text) because that's what we have in the sidebar/stream, and it was specifically requested in those places because it was useful info for some documents. It's not in Conor's mockup though, and it does sometimes wrap onto a new line.

I also added an arbitrary amount of vertical space between `<li>s` to make it a little more readable.

seanh commented:
Tests added and rebased

robertknight commented:
The initial view after you've just created a new group looks somewhat bare and the most important thing we want users to do next, share the group, is tucked down at the bottom of the page.

@conordelahunty - Thoughts?

![empty-group-view](https://cloud.githubusercontent.com/assets/2458/10756232/92d1070a-7c9a-11e5-8b98-5e7c8f6d1016.png)

robertknight commented:
@seanh Couple of UX notes:

* I'm not sure if was on the design, but I generally avoid contractions ("hasn't") in UI text
* The link should be focused when the page loads so the user can just press Ctrl+C to copy or tap on mobile

seanh commented:
> I'm not sure if was on the design, but I generally avoid contractions ("hasn't") in UI text

I generally use them on purpose! Makes it more casual/friendly imo. But afaik we don't have any style guide for this stuff. We need one. Conor's not here this week either so we can't really defer to him on it. In this case I'll just go ahead and change it

seanh commented:
> The link should be focused when the page loads so the user can just press Ctrl+C to copy or tap on mobile

This already happens for me in Chrome and Firefox. Is it not happening for you?

seanh commented:
Changed hasn't -> has and rebased on latest master.

Shall I go ahead and add 1em of left and right padding to `.group-form`? I think it needs it, as some document titles go right up to the very edge of the white box

robertknight commented:
> Shall I go ahead and add 1em of left and right padding to .group-form?
> I think it needs it, as some document titles go right up to the very edge of the white box

I'm using `px` measurements for everything new as it is much easier to match up directly with the designs.

robertknight commented:
I can see room for a bunch of design tweaks but I think it is important to get this landed and give the others a chance to play around with it.

robertknight commented on h/api/models.py:
```diff
@@ -166,6 +170,112 @@ def title(self):
+        """
+        uri = jinja2.escape(self.uri) or ""
+
+        if uri.startswith("http://") or uri.startswith("https://"):
```
Is the uri normalized here so that the scheme is always lower-case?

seanh replied:
It's not, just pushed a fix and tests for this

seanh commented:
> > Shall I go ahead and add 1em of left and right padding to .group-form?
> > I think it needs it, as some document titles go right up to the very edge of the white box
> 
> I'm using `px` measurements for everything new as it is much easier to match up directly with the
> designs.

10px?

seanh commented:
> I can see room for a bunch of design tweaks but I think it is important to get this landed and give the others a chance to play around with it.

Agreed

robertknight commented on h/api/models.py:
```diff
@@ -166,6 +170,112 @@ def title(self):
+                domain = urlparse.urlparse(uri).hostname
+        else:
+            if uri.startswith("file://"):
+                title = urllib2.unquote(jinja2.escape(self.filename))
```
I'd strongly advise not mixing up logic for determining field values with logic for performing HTML escaping.

There is an HTML injection vulnerability here:

```
1. Create a file with the title `"><script>window.alert("Pwned");<%2Fscript>.html` and no content (especially, no <title>)
2. Service this via a local HTTP server and annotate it in a group
3. Refresh groups page
```

robertknight replied:
The way I would approach this is to split the rendering into two parts - one which determines the different parts of the link (href, title etc.) and another which renders the appropriate link, escaping the values accordingly.

seanh replied:
Is that injection really possible? I'm escaping the title, and I have a test for it

seanh replied:
> The way I would approach this is to split the rendering into two parts - one which determines the different parts of the link (href, title etc.) and another which renders the appropriate link, escaping the values accordingly.

It's probably a good idea to break this up, as it's a long function

seanh replied:
This has been broken up here https://github.com/hypothesis/h/pull/2683

robertknight commented on h/api/search/query.py:
```diff
@@ -96,8 +96,18 @@ class AuthFilter(object):
+        self.private = private
 
     def __call__(self, params):
         groups = list(self.request.effective_principals)
```
This initially confused me because 'groups' is not in fact, just a list of groups but a set of `principals` which can include both groups and user IDs. I suggest renaming the `groups` variable here accordingly (eg. to `principals`)

seanh replied:
Renamed

robertknight commented on h/api/search/test/core_test.py:
```diff
@@ -0,0 +1,30 @@
+
+    core.search(request, mock.Mock(), private=True)
+
+    query.AuthFilter.assert_called_once_with(request, True)
```
Could this be written as `assert_called_once_with(request, private=True)`? If so, that would be clearer.

seanh replied:
Done

robertknight commented on h/api/test/models_test.py:
```diff
@@ -78,6 +79,108 @@ def test_title_annotation_that_has_no_document():
+        uri='http://example.com/example.html',
+        document={'title': '</a>' + spam_link})
+
+    assert spam_link not in annotation.document_link
```
This doesn't verify that the link was escaped properly, only that the injected link doesn't occur _exactly_ as in the input text. I'd suggest checking for the presence of any escaped chars - eg. by checking that `escape(input) == input`

seanh replied:
This checking is done differently in [the refactor](https://github.com/hypothesis/h/pull/2683)

robertknight commented on h/static/styles/group-form.scss:
```diff
@@ -73,6 +74,14 @@
+  }
+
+  .group-form-document-list li {
+    margin-top: .5em;
```
I'm generally preferring `px` units in new styling as this makes it easier to ensure fidelity with mockups.

nickstenning commented on h/api/models.py:
```diff
@@ -166,6 +175,113 @@ def title(self):
+            return ""
+
+    @property
+    def document_link(self):
```
This is starting to look increasingly like presenter code, and it makes me a bit uneasy to have it in the model. I think for now we merge this as it is, but in future I think we should perhaps consider having an `AnnotationPresenter` class which takes an `Annotation` in its constructor and is responsible for all the mapping between the data model and the presentation model. This would presumably also make the logic in these things a bunch easier to test as it wouldn't require setting up the database connection, etc.

seanh replied:
Do I need to setup the database model to test these?

But yes we could move it out of the model. If we did so it might be convenient if `search.search()` returned `AnnotationPresenter` objects, or maybe it's better if user code (such as view code) wraps `Annotation`s in `AnnotationPresenter`s itself.

nickstenning commented on h/api/models.py:
```diff
@@ -166,6 +175,113 @@ def title(self):
+        if domain:
+            domain = truncate(domain)
+
+        assert title, "The title should never be empty"
```
Perhaps it's just a matter of preference, but I'd rather see the code structured in a way that makes it self-evident that the title is not empty, rather than using assertions like this.

That is, flip the logic on its head and take a "progressive enhancement" approach rather than the "graceful degradation" one you're employing here:

```
uri = jinja2.escape(self.uri) or ""
title = urllib2.unquote(uri)

if self.title:
    title = ...
elif is_file(uri):
    title = ...
```

seanh replied:
I think these are gone in [the refactor](https://github.com/hypothesis/h/pull/2683)

nickstenning commented on h/api/models.py:
```diff
@@ -166,6 +175,113 @@ def title(self):
+            link = '<a title="{title}">{link_text}</a>'.format(
+                title=title, link_text=link_text)
+        else:
+            assert False, "We should never get here"
```
This assertion is unnecessary if you just reduce the `if` clause to its minimal form.

```python
if href and domain:
    ...
elif domain: # domain and no href
    ...
elif href: # href and no domain
    ...
else: # no domain and no href
    ...
```

seanh replied:
Also gone in [the refactor](https://github.com/hypothesis/h/pull/2683)

nickstenning commented on h/api/search/test/core_test.py:
```diff
@@ -0,0 +1,30 @@
+@pytest.fixture
+def query(request):
+    patcher = mock.patch('h.api.search.core.query', autospec=True)
+    request.addfinalizer(patcher.stop)
```
Small request. Can you make sure not to add the finalizer until the patcher is actually started? Otherwise a failure to patch (perhaps because the module moved) will also spew an error about "called stop on unstarted patcher" and the real error gets hidden further up. So:

```python
patcher = mock.patch(...)
result = patcher.start()
request.addfinalizer(patcher.stop)
return result
```

seanh replied:
To be done here https://github.com/hypothesis/h/pull/2683

nickstenning commented on h/groups/templates/create.html.jinja2:
```diff
@@ -18,8 +18,7 @@
       <div class="group-form-footer__explain-text js-group-info-text is-hidden">
-        <p class="group-form-footer__heading">About Groups</p>
-        <p>Groups let you annotate and discuss content on the web privately.</p>
+        {% include "about-groups.html" %}
```
This file should have a `.jinja2` extension.

seanh replied:
Done on [the other pull request](https://github.com/hypothesis/h/pull/2683)

nickstenning commented on h/groups/test/views_test.py:
```diff
@@ -349,6 +349,125 @@ def test_read_if_already_a_member_returns_response(Group, renderers):
 
 
+@read_fixtures
+def test_read_calls_search(Group, search, renderers):
```
Small quibble on the name. We're not testing that it calls `search`. We're testing the collaboration between the view function and the `search` function -- that is, we're testing that it calls `search` with the correct arguments. Might be nice to make that clearer in the method name or the docstring, or both.

seanh replied:
Done

nickstenning commented on h/groups/test/views_test.py:
```diff
@@ -349,6 +349,125 @@ def test_read_if_already_a_member_returns_response(Group, renderers):
+
+@read_fixtures
+def test_read_calls_normalize(Group, search, renderers, uri):
+    """It shold call normalize() with each of the URIs from search()."""
```
I'm not really sure about this test. Again, what I think you actually want to test is that if `search` returns N annotations whose URIs differ only in normalisation, you only get one back, no? There are many ways of doing this, some of which don't involve calling `uri.normalize` (e.g. partition on `target.source` rather than `uri.normalize(annotation.uri)`).

seanh replied:
This is literally just supposed to be testing that it calls `normalize()` once for each of the URIs that `search()` returns. But I think you're right that the test below (the one that tests that duplicate normalized URIs are removed) covers this, so this one can just be deleted

nickstenning commented on h/groups/test/views_test.py:
```diff
@@ -349,6 +349,125 @@ def test_read_if_already_a_member_returns_response(Group, renderers):
+
+
+@read_fixtures
+def test_read_duplicate_documents_are_removed(Group, search, renderers, uri):
```
So yes, if you've got this test, then the test I mentioned in [this comment](https://github.com/hypothesis/h/pull/2667/files#r43128239) seems redundant and overly brittle.

nickstenning commented on h/groups/test/views_test.py:
```diff
@@ -349,6 +349,125 @@ def test_read_if_already_a_member_returns_response(Group, renderers):
+    search.search.return_value = {"rows": annotations}
+
+    def normalize(uri):
+        # All three annotations' URIs normalize to the same URI.
```
`s/three//`

nickstenning commented on h/groups/views.py:
```diff
@@ -101,7 +105,28 @@ def _read_group(request, group):
+    # Group the annotations by URI.
+    # Create a dict mapping the (normalized) URIs of the annotated documents
+    # to the most recent annotation of each document.
+    annotations_by_uri = collections.OrderedDict()
```
I'm not sure if you'd say this is easier or harder to read, but it makes a bit more sense to me:
```
buckets = itertools.groupby(annotations, lambda a: uri.normalize(a.uri))
document_links = itertools.islice(next(anns).document_link for _, anns in buckets), 25)
```

seanh replied:
Matter of opinion I guess, I find my way easier to read. Mine also breaks as soon as it has 25 unique normalized URIs.

nickstenning commented:
I'll also note that we should probably remove the autofocus from the group share field as part of this work, as this causes the viewport to jump down the page on load when there are more than a few "recently annotated documents."

nickstenning commented:
I'm going to merge this as-is, @seanh. It works, and it can serve as the basis for some design iteration. I would like it if you could go through the remaining comments on this nonetheless. Cheers.

seanh commented:
Unfortunately testing this will be a little pointless as I'm about to push a complete refactor of the document link rendering :(

seanh commented:
Continuing here https://github.com/hypothesis/h/pull/2683

seanh commented:
Ok, I've gone through the comments, I think they've all been replied to here or else done on <https://github.com/hypothesis/h/pull/2683>, a few are marked as TODOs on <https://github.com/hypothesis/h/pull/2683>, will continue this work over on <https://github.com/hypothesis/h/pull/2683>.
